### ISSUE TITLE
Make classes final by default

### ISSUE THREAD

**Issue by ciaranmcnulty** [phpspec/phpspec#632]
This would be a fairly simple template change.

Would it kill our user base? :-)

/cc @everzet

**Comment by ciaranmcnulty** [comment]
Pro: Promotes something that I'd consider a best practice

Con: Might freak out some newcomers (this isn't that important)

**Comment by stof** [comment]
should we make this configurable (without having to overwrite the whole template, which is kind of hard as it forces to update it for any improvement done in PhpSpec later) ? This way, we could satisfy both people wanting final classes and people not liking them

**Comment by ciaranmcnulty** [comment]
We could; we could have two templates and switch between them based on config.

@mdavis1982 had a PR that implemented this at one stage.

I'm not sure about having two ways to configure it though, and if we do it it should default to 'ON'

**Comment by mdavis1982** [comment]
I still have the code for this PR lying around... At the time, I didn't know how to add the tests needed for the feature to be included. Now that I do, I'll try to resurrect it ASAP

**Comment by MarcelloDuarte** [comment]
+1 for configurable, not default behaviour

**Comment by ciaranmcnulty** [comment]
default on or default off, though?

I mean, it's configurable right now by making a personal class template in `~`

**Comment by norzechowicz** [comment]
Maybe it could be a question before generating class with default value "yes"?
Or maybe class could be generated as finall when there will be something like $this->shouldBeAFinal() when class does not exists yet?

**Comment by MarcelloDuarte** [comment]
I suggest that by default it is off, but a simple flag in phpspec.yml would turn the final template on.

**Comment by ciaranmcnulty** [comment]
If the default is off, I don't see any benefit in having the option - if a user is savvy enough to switch it on in phpspec.yml they're savvy enough to edit the template.

**Comment by mdavis1982** [comment]
@ciaranmcnulty I've just open the above PR to address this.

**Comment by stof** [comment]
@ciaranmcnulty overwriting the template means that they need to sync their custom template with changes done in phpspec all the time. It is less maintainable than a flag in phpspec.yml

**Comment by MarcelloDuarte** [comment]
Here's an idea, if `shouldHaveType` or any [type matcher](http://www.phpspec.net/en/latest/cookbook/matchers.html#type-matcher) is used, and the interface does not exist, then offer to create the interface AND ask if you want the class to be final. (Inspired by @Ocramius [blog post](http://ocramius.github.io/blog/when-to-declare-classes-final/))

**Comment by ciaranmcnulty** [comment]
Generating the interface is something that came up in the past, it's worth considering in a different issue. The `final` thing would be a nice option in that.

**Comment by ciaranmcnulty** [comment]
@stof I'm not sure the class template iterates quickly enough to worry about people having to update their templates. Anyhow, this change would require everyone to update their custom templates to look at the switch, right?

**Comment by mdavis1982** [comment]
@ciaranmcnulty I don't think they'd need to update their templates... If their templates don't include it then it will just be ignored?

**Comment by ciaranmcnulty** [comment]
I think #633 looks pretty good. I've no strong objection to adding it aside from questioning whether it's that useful given users can edit the templates themselves.

@mdavis1982 I meant that they'd have to edit their custom template to benefit from this new flag/placeholder

@MarcelloDuarte This idea's grown on me, I'm not sure it excludes the suggested finalise-everything config option

**Comment by everzet** [comment]
I'm highly against providing such options. Either we believe this is a good design decision by default or we don't. Throwing choice on the users hardly will solve anything.

**Comment by jakzal** [comment]
I'm :+1: to generate final classes by default. 

However, I also think we should start documenting such decisions with explanations (or links to explanations) on why we claim these are good things.

**Comment by mdavis1982** [comment]
@everzet I think I agree with you. Personally, I think this option should be enabled by default and it should be disabled explicitly if the user doesn't want the behaviour. I'd like the option to still be there, but I think enabling it by default would be the best idea.

**Comment by nchagrass** [comment]
Not everyone is familiar with the idea of having its classes final until you need to change it.

Who is your audience? who are your users? are they interested in having this? does they want it ?

Make sure they are confident with this concept, prepare a well made documentation ; )

**Comment by richardmiller** [comment]
Documenting these things sounds good. 

Could have it on with no option to turn it off but document how to stop it generating final by providing custom templates. Then people can turn it off if they must but it will require some effort so it's more of a proper decision that just not liking it and immediately turning it off.

**Comment by mdavis1982** [comment]
@richardmiller That would work... But I think, as a user, I'd prefer to be able to turn it off without having to put stuff in my home directory. I might want it on for one project, and off for another, for example.

**Comment by bestform** [comment]
I don't think this should be something that is on by default. People who know their stuff will either change the template or update the generated class afterwards. People who still struggle with OOP will be very confused by this. 

I also second @everzet 's quote: "Delegating choice to the users hardly helps anyone ever."

**Comment by everzet** [comment]
> People who still struggle with OOP will be very confused by this.
May I remind everyone that exposing people's struggle with OOP is actually one of the key goals behind PhpSpec? :)

**Comment by jakzal** [comment]
>  People who still struggle with OOP will be very confused by this. 

People who struggle with OOP will struggle with PhpSpec too. There's already [loads of reasons for it](http://techportal.inviqa.com/2014/09/11/my-top-ten-favourite-phpspec-limitations/). 

> I also second @everzet 's quote: "Delegating choice to the users hardly helps anyone ever."

That's exactly why we should just make this option a default ;)

**Comment by mdavis1982** [comment]
@jakzal So you're in favour of having it on by default, and allowing the user to switch it off with an option in `phpspec.yml`?

**Comment by jakzal** [comment]
@mdavis1982 no, we don't need an option. If we keep adding options for each little feature, we'll end up with sh*t loads of switches ;)

I think it's fine to have it as a default behaviour. It's not a big deal to remove the final keywords after a class was generated. Most people will leave it in, which is good - it'll communicate that the author haven't really thought of this class as an extension point.

**Comment by icambridge** [comment]
My question is what problem are you trying to solve here? It seems like this is just a matter of opinion?

**Comment by mdavis1982** [comment]
@jakzal I disagree...

> I think it's fine to have it as a default behaviour. It's not a big deal to remove the final keywords after a class was generated. 

It's also not a big deal to add them afterwards. :wink: I'm joking, of course - but I'm not convinced that's a good enough reason to remove the option.

I think it's important to be able to disable it... If I'm working on a project where `final` classes aren't used, I then have to go and remove it *every* time something gets generated. We could argue that the classes *should* be `final`, but the reality is that the tool is probably used in places where that's not going to be happening any time soon.

I can't just override the templates to remove it, because if I'm working on more than one project (which is the case for me at the moment - one of which uses final and one that doesn't), then that'll change the templates for all projects - so I have to go and edit the code for one of them whatever if we don't add the option.

**Comment by everzet** [comment]
@icambridge problem is people abusing inheritance. Enforcing `final` classes by default introduces explicit cognitive step before extending classes and based on our experience effectively helps people to prefer composition over inheritance.

**Comment by jakzal** [comment]
@mdavis1982 in which case I'd see project specific templates as a better solution

**Comment by mdavis1982** [comment]
@jakzal I'm inclined to agree - mostly.

I think for something like this, overriding the entire template seems a bit overkill.

**Comment by mlopes** [comment]
> if I'm working on more than one project (which is the case for me at the moment - one of which uses final and one that doesn't), then that'll change the templates for all projects - so I have to go and edit the code for one of them whatever if we don't add the option.

Doesn't this already happen? The only change is that instead of adding it to the project that uses final, you'll need to remove it from the one that doesn't.

**Comment by mdavis1982** [comment]
@mlopes That is what happens at the moment. If we add the option to `phpspec.yml` then one project will generate `final` classes by default, and the other won't. I won't have to edit the code at all.

**Comment by rtuin** [comment]
Great idea, good discussion. I support generating final classes.

Some thoughts:
* Should be default behaviour
* User should not be able to turn this off via configuration, but must change the template instead (or remove the `final` keyword on generated classes)
* As this encourages design by contract the user could be asked to generate a belonging interface (with e.g. `<concretename>Interface` as name). I feel this will improve the workflow for users.
* Introduce this only in a new major version
* Reason behind this should be clear from the (getting started) docs.

**Comment by ciaranmcnulty** [comment]
@rtuin Being prompted to create an interface for every class would fast become a living nightmare, and calling it `*Interface` is something I'd actively discourage.

**Comment by ciaranmcnulty** [comment]
@mdavis1982 You can have project-specific templates; I don't see that as more overhead than editing phpspec.yml.

Actually this project should have a template with the copyright notice in it maybe.

**Comment by mdavis1982** [comment]
@ciaranmcnulty I wasn't aware of that... Maybe I should have read the docs more closely! :wink: 

I still think that being able to turn it off via a switch in the `phpspec.yml` is easier for the user - and it doesn't require them to have to go find the template from the `vendor` folder to change a single word in the template. It just seems a bit overkill to have to go to all that trouble for this.

**Comment by rtuin** [comment]
> Being prompted to create an interface for every class would fast become a living nightmare, and calling it *Interface is something I'd actively discourage.

@ciaranmcnulty I was just thinking out loud and I agree with you. We could perhaps generate the Interface from the `shouldImplement()` method, but that's a whole different feature. Not worth discussing further in here.

**Comment by ciaranmcnulty** [comment]
@rtuin Yeah I think that's a good feature, offering to create interfaces from `shouldImplement` or `shouldHaveType` and it's definitely on the roadmap.  Above @MarcelloDuarte suggests at that point also offering to finalise the class, which is interesting.

**Comment by mdavis1982** [comment]
@ciaranmcnulty It's interesting - but adds an extra question to answer which I'm not sure is worth it... They could just set the option and then it wouldn't have to bother them :wink:

**Comment by rtuin** [comment]
@ciaranmcnulty @MarcelloDuarte That's a great idea! But I agree with @everzet that we shouldn't give users too many choices. So how about the final keyword is automatically added after the generation triggered by a `shouldHaveType()` or `shouldImplement()`. If I'm not mistaken the entire suite is run after code generation, so this will instantly reveal the possible design flaws/inheritance abuse to the user.

**Comment by mheki** [comment]
:+1: for the `final` plus documentation explaining why and how to change it.
As @jakzal said - shed loads of options would complicate clean phpspec's API.

**Comment by stof** [comment]
I opened an issue about explaining such decisions in the doc (not only for final)

**Comment by ciaranmcnulty** [comment]
Thanks @stof, that's a good idea.

**Comment by mathiasverraes** [comment]
- +1 on final ([obviously](http://verraes.net/2014/05/final-classes-in-php/))
- +1 on default, not configurable. 

PhpSpec pushes people to better design. The question whether users will like it or not, is not part of the discussion on design. It is a marketing and documentation problem (which is often the same thing in OSS).  Just be prepared to explain the reasoning behind it over and over. I know I have :-)

**Comment by mdavis1982** [comment]
I've opened another pull request that will generate `final` classes by default, without making it configurable via the `phpspec.yml` file.

**Comment by MarcoDeBortoli** [comment]
* +1 on final
* +1 on default, not configurable but templatable
* -1 *Interface suffix
* +1 `shouldImplement()` for interface generation

**Comment by MarcelloDuarte** [comment]
* +1 on final, default for classes with interfaces mentioned in shouldHaveType(name different to SUS)
 * -1 on prompting to create an interface for every class
 * -1 *Interface suffix (eew)

**Comment by mdavis1982** [comment]
- +1 on final
- +1 on default, not configurable but can be overridden in templates
- -1 *Interface suffix

**Comment by mdavis1982** [comment]
When is GitHub going to start counting :+1: and :-1: in issues?!

**Comment by ciaranmcnulty** [comment]
It's not a democracy :-P

**Comment by ciaranmcnulty** [comment]
OK, I was fairly convinced before, I'm even more convinced now.

This isn't going to happen on 2.2, not because of any difficulty in implementing it but because we do need to get serious about documentation to support these sorts of opinionated decisions and help educate users. This doesn't fit with 2.2 being released soon.

I have one reservation, which is that final classes will push our users to use real objects inside specs instead of doubles. This is appropriate for values and other trivial objects, but we need to help users see those distinctions. There are some known issues where users pass real-objects-containing-doubles into their SUS and some focus on them would help.

Rapidly after the 2.2 release I'll make final the default in master and we go from there.

**Comment by mdavis1982** [comment]
@ciaranmcnulty Do you need my PR?

**Comment by ciaranmcnulty** [comment]
@mdavis1982 I'll leave it open if that's ok.

**Comment by mdavis1982** [comment]
@ciaranmcnulty Yep, no problem at all. Shall we close the one that adds the option?

**Comment by ciaranmcnulty** [comment]
Update on this, I have a kind-of roadmap:

- [X] Make PhpSpec's own classes final going forward (see #713 and https://github.com/phpspec/phpspec/wiki#coding-standards)
- [] Retrospectively make existing PhpSpec classes final where possible as changes are applied to them (this has to be done thoughtfully with reference to https://github.com/phpspec/phpspec/wiki#versioning-policy)
- [] Make the default user template final (as in #637) - this will require documentation on why the decision is made, and how to turn it off; probably also some advice on how to test interactions between final objects.
- [] Finalize all relevant internal classes as part of `3.0.0`

**Comment by icambridge** [comment]
Just thinking wouldn't make a lot of peoples code instantly breach the open/close principal or do you think people who care about that would know to remove final? Or am I missing something?

**Comment by ciaranmcnulty** [comment]
If they're not already type hinting against arrays or abstracts, making the concrete classes final won't affect their ability to reuse code too much.

**Comment by icambridge** [comment]
Well Open/Close principal is about being open to extension while closed to modification no? final instantly removes the ability to extend? For example I will no longer be able to extend PHPSpec for whatever reason, I would have to fork it and maintain my own copy. Obviously not a biggie since it's configurable.

**Comment by ciaranmcnulty** [comment]
You can replace any of the classes with your own implementation that conforms to the interfaces (this is what Behat v3 does and it works well)

**Comment by icambridge** [comment]
That's because of Dependency Inversion which is different to Open/close no? Open/close is on classes and stuff no? Say there is a class in PHPSpec that does 90% of what I want I just want to add in something on top. Instead of being able to create my class by extending I have to copy and paste the class or fork the entire project just to remove the keyword final. Either option doesn't sound too fun.

**Comment by ciaranmcnulty** [comment]
You can use a Decorator, or you can PR to the project to add an extension point. 

Most of the time these things indicate a design problem with the original classes.

**Comment by icambridge** [comment]
Say https://github.com/phpspec/phpspec/blob/master/src/PhpSpec/CodeGenerator/GeneratorManager.php had a final keyword.

I wanted to extend it so I could get all generators. Other than being blocked because it's got a private keyword. I would also be blocked because of the final keyword. There is no interface (I assume rightfully so because there is no need for one) and even if there was I still wouldn't be able to access the protected member to allow me extend it. Not saying there is a need in the exact use case. More than wouldn't final by default be breaching open/close? I've been thinking a lot about open/close lately...

**Comment by ciaranmcnulty** [comment]
I really don't think the open/closed principle is about extending classes to change the parts of them you don't like.

In the example you give it sounds like you want to extend the class so that you can add a getter to break encapsulation? I don't care about that use case.

**Comment by stof** [comment]
the open/close principle is about providing extension points. This does **not** mean allowing inheritance on the class. Inheritance is most of the time the worst way to provide extension points for a library.

**Comment by jakzal** [comment]
> I really don't think the open/closed principle is about extending classes to change the parts of them you don't like.

:+1: 

@icambridge the example you gave is a perfect example of the Open/Closed principle. Making the GeneratorManager final doesn't make any difference, as it's not meant to be extended (therefore there's no interface). However, extension point is there, as you're able to provide your own generators by implementing the GeneratorInterface. This is what OCP is about.

**Comment by ciaranmcnulty** [comment]
A good example is #682 

The initial implementation was to change a `private` to a `protected` and achieve an extension point via inheritance. Through discussion we realised there was a missing concept that's been turned into a collaborator and made explicit. 

This has hugely simplified things for MageSpec, made PhpSpec better, and will probably help future extensions to do things cleanly.

If we'd started off with a `protected` method it could have been lazily extended and replaced.

Making the class `final` will help the same way having things `private` does; few developers argue for `protected` over `private`, but PHP devs just aren't used to `final` yet.

**Comment by ciaranmcnulty** [comment]
I'm going to edit the checklist above into the first comment

**Comment by icambridge** [comment]
> I really don't think the open/closed principle is about extending classes to change the parts of them you don't like.

@ciaranmcnulty  In the example I gave I wasn't changing anything, I was merely adding more things. 

Open/Close is directly related to not changing the original class to add new features? No? 

> the open/close principle is about providing extension points

@stof are you sure? http://www.objectmentor.com/resources/articles/ocp.pdf directly talks about inheritance and making things protected and public instead of private. I know this isn't from the original author of Open/Close and things change but it's still I think a pretty good resource for the definition. 

@jakzal I know that I constantly think I'll build something for one thing and I'll never need to do anything else and then someone else comes along with another task and they did need to do stuff with that code for valid reasons. Just because I never meant for it to be used for something else doesn't mean someone else hasn't seen it another use for it to solve another problem. 

Also isn't the changes that are available to be made there because of Dependency Inversion and good design rather than obeying the Open/Closed principal.

**Comment by ciaranmcnulty** [comment]
"Open/Close is directly related to not changing the original class to add new features"

It's about being able to use a class for new things without adding features

As far as I can see all of your examples can be achieved with composition solutions rather than inheritance.

**Comment by Ocramius** [comment]
> making things protected and public instead of private

That's why most of the classes in ZF2 are doomed. It's a mess, don't do that, ever :-(

**Comment by ciaranmcnulty** [comment]
It also prevents us making some changes that would help PhpSpec grow

**Comment by icambridge** [comment]
> As far as I can see all of your examples can be achieved with composition solutions rather than inheritance.

The first example couldn't be, since there is no interface. To apply the interface would result in a change to the original source code which Open/Closed is designed to stop?

I'll stop now... :smile:

**Comment by ciaranmcnulty** [comment]
'I want to get all the generators' => get them from the place that passes them to GeneratorManager

**Comment by icambridge** [comment]
Ok I lied about stopping I can't help myself...

> 'I want to get all the generators' => get them from the place that passes them to GeneratorManager

For whatever reason the generators are generated in various places, there is no single point. Also I don't want to modify the original code that creates them?

**Comment by ciaranmcnulty** [comment]
@icambridge The magic happens here https://github.com/phpspec/phpspec/blob/master/src/PhpSpec/Console/ContainerAssembler.php#L200-L209

It's trivial to get the same list and inject somewhere else. Plus this use case is totally made up :-)

**Comment by ciaranmcnulty** [comment]
Possibly if there was a real use case we'd introduce a GeneratorManager interface

**Comment by icambridge** [comment]
@ciaranmcnulty I think you're thinking too closely to the application, being open/close instead of the units of code. For example say for whatever reason I somehow discovered that the perfect solution was to reuse phpspec as library for my own testing/code generation thing. I have a UI I want to display all the generators available. I wouldn't be able to use an extended GeneratorManager to do that. I would have to break DRY and create another class to hold all these Generators. While improbable it's not impossible that could happen. So improbable I wouldn't go out my way to change things, but still makes me wonder about Open/Close. As it seems there is no way to extend that class. It's closed from extension and from modification.

**Comment by MarcoDeBortoli** [comment]
@icambridge I think the problem here is that while it's true that based on OCP software entities should be open for extension, but closed for modification, it's also true that based on DIP one should depend upon Abstractions not upon concretions.
As @ciaranmcnulty pointed out "if there was a real use case we'd introduce a GeneratorManager interface".

I do generally try to look at SOLID principles as a whole rather than picking one of each situationally.

**Comment by quimcalpe** [comment]
@icambridge I happily break DRY when is worth it

**Comment by icambridge** [comment]
@MarcoDeBortoli Having an interface doesn't make GeneratorManager extendable, that class is simply not extendable. It has a single responsibility of providing generators upon request, adding the functionality of returning them all for whatever reason wouldn't be breaking that. Adding that functionality wouldn't break the original functionality so Liskov's principal wouldn't be broken, there is no interface so interface segregation isn't broken. The method wouldn't be breaking dependency inversion.  

Also adding interfaces doesn't make it OCP friendly, it means I can add can modify the existing functionality that is there by decorators but it still doesn't allow me to add this getAllGenerators I seem to want so much.

@quimcalpe but this wouldn't be you choosing to, this would be it forced upon you because something simply wasn't open to being extended.

**Comment by ciaranmcnulty** [comment]
From a discussion with @everzet there's a softer option where we make it easier to generate classes based on interfaces (perhaps using `shouldHaveType`) and make them final by default

**Comment by MarcoDeBortoli** [comment]
@icambridge it will actually do break DIP because you are about to create in your code a dependency on a method that it's not part of a well defined contract but it's consequence of a specific personal need. I think it has to be considered that PhpSpec as a library has its own set of rules and design decisions and their maintainers are designing it with a specific and precise idea and goal in mind. 

My position is that your requirement is somehow beyond PhpSpec responsibility, if you really have the need of extend the GeneratorManager as of now fork is the only solution or maybe you can just simply proxy it. I'm saying your requirement but what I really mean is anyone else requirment. If @ciaranmcnulty and the other guys have to shape PhpSpec based on everyone exception it will be hell and dragons.

That said, it doesn't mean that this will never happen. The adoption of final has been widely discussed and the community reached a sort of consensus but if a real need of extensibility for the generator manager becomes core surely the PhpSpec team will not object in make it dependent on an interface.

I don't know, @ciaranmcnulty is probably the best one to answer rather than me at this point

**Comment by ciaranmcnulty** [comment]
Sure, if someone had a valid use case for replacing the GeneratorManager we can add an interface.

A whole load of  BC headaches come from concrete implementations being extended elsewhere.

**Comment by icambridge** [comment]
Guys I am not suggesting that we change GeneratorManager or anything for that matter. I was literally using that as an example class since it had no interface.  I am merely pointing out that if I was to use PHPSpec as a library to build on top of I couldn't extend that one class I would have to copy and paste it and then add that single feature. Maybe there is a use case for that, that no one knows yet. That would be why OCP exists.

> My position is that your requirement is somehow beyond PhpSpec responsibility, if you really have the need of extend the GeneratorManager as of now fork is the only solution or maybe you can just simply proxy it. 

That's exactly my point and why I don't think that class is OCP. It's just come to my attention that OCP is some how ignored misinterpreted in my opinion to mean if I use all the other principals OCP is suddenly happens.

And as I said doesn't really matter in the long run since this option is configurable.

**Comment by MarcoDeBortoli** [comment]
It actually does matter. As a community we should all collaborate to improve our tools and I think you had a good shout. You saw something that could be a problem so why not try to see if we can to better if needed :)

**Comment by jfmercer** [comment]
\<my-two-cents\>

@ciaranmcnulty New guy here (i.e., 1 year of PHP experience). I use PHPSpec because it forces me to follow best practices that I don't even know exist. Basically, when I want to write a test but can't do it in PHPSpec, I know that the problem is me, not the test framework. If making classes final is a best practice, then, IMHO, it should be the default behavior. 

PHPSpec is not merely a testing framework; it is also a pedagogical tool. It should default to best practices and let its users "catch up" to it, so to speak.

\</my-two-cents\>

**Comment by MarcoDeBortoli** [comment]
^ that

**Comment by ciaranmcnulty** [comment]
#713 is merged, it will be interesting to see whether this makes development of PhpSpec easier or harder

**Comment by icambridge** [comment]
@jfmercer on a separate point but related to your comment. I think one thing I think we should always ask and that is what is Best Practice. What was best practice sudden isn't. Like it used to be the thing to use magic methods all over the place, now people don't do it. I personally don't think you should ever take one persons (or in this case lots of other people) word on if something is best practice or not. You should read the well defined tried and tested principals and then make your own decision.  

Relying on PHPSpec to tell you what is best practice can result in your doing the wrong thing simply because everyone else thought it was a good idea for some reason. Just try not to follow in to the Groupthink trap. Not saying it's happening but it's possible.

**Comment by jfmercer** [comment]
@icambridge I was not suggesting that PhpSpec is _The Ultimate Judge of Best Practices Forever™_. Nor would I suggest that beginners should rely on it exclusively to learn best practices. However, it is an opinionated test framework, and this is one of its greatest strengths. It forces its users to write code _in a particular way_ in order for it to be testable. In contrast, PhpUnit can test almost anything, ever the worst of code. Of course, that it just as it should be, given that the premises and goals of PhpUnit are quite different from those of PhpSpec. But this flexibility comes at a price: PhpUnit is a poor pedagogical tool, as whereas PhpSpec is.

It takes years of hard work, failure, and mentoring in order to understand how to write good, clean OO code. PhpSpec forces its users to write code well, much as a piano teacher forces his students to sit and breath and play in a certain way. Of course, once the student masters the piano, he may have very different opinions and styles than his former mentor. Nevertheless, the fully mature professional would never have achieved mastery without the strong opinions of his tutor. The same is true in programming: a beginner should follow the guidance of masters until he becomes a master himself. Only then does his programming style and philosophy bear any weight or authority. 

It is obvious that PhpSpec is a testing framework. My comment above emphasizes that, however imperfect, it is _also_ a pedagogical tool. If PhpSpec were to abandon the enforcement of best practices, and leave it to the user to enforce these practices himself in a config file, then PhpSpec might remain a good BDD framework. Perhaps it would be more successful, gain a wider audience, and even became the PHP industry-standard for testing. But it would cease to be a teacher; it would no longer have any pedagogical value. 

My suggestion is that, amongst their other priorities, PhpSpec's developers should keep beginners in mind. As this framework becomes less and less opinionated, it will become less and less useful to apprentices in the craft of programming.

**Comment by ciaranmcnulty** [comment]
Thanks for the kind words @jfmercer 

It's good to hear this perspective, we quite often are scared to implement changes in case 'it scares of newcomers'.

**Comment by MarcoDeBortoli** [comment]
I second @jfmercer word by word.

**Comment by icambridge** [comment]
@jfmercer My point was more related to not just listening to what others think to decide what you think is best practice. What makes it best practice? Questions should be asked such as do other programming languages that support this feature also do it like this if not why not? Etc. My comment was more general advice and not really related to PHPSpec.
